rasa/shared/utils/io.py

```diff
@@ -318,25 +319,99 @@ def env_var_constructor(loader, node):
+            replace_environment_variables()
+        else:
+            # we need to do this because `yaml` adds contains a global `implicit_resolvers` list
+            # and the resolves from there cannot be de-registered
+            yaml.SafeConstructor.add_constructor("!env_var", lambda _, node: node.value)
+
+        self._save_modified_yaml_parameters()
+        self.restore_default_yaml_parameters()
+
+    def _save_default_yaml_parameters(self) -> None:
+        self._default_yaml_implicit_resolvers = (
+            yaml.Resolver.yaml_implicit_resolvers.copy()
+        )
+        self._default_yaml_constructors = yaml.SafeConstructor.yaml_constructors.copy()
+
+    def restore_default_yaml_parameters(self) -> None:
+        """Restores the `ruamel.yaml` parameters that were specified before."""
+        yaml.Resolver.yaml_implicit_resolvers = self._default_yaml_implicit_resolvers
+        yaml.SafeConstructor.yaml_constructors = self._default_yaml_constructors
+
+    def _save_modified_yaml_parameters(self) -> None:
+        self._modified_yaml_implicit_resolvers = (
+            yaml.Resolver.yaml_implicit_resolvers.copy()
+        )
+        self._modified_yaml_constructors = yaml.SafeConstructor.yaml_constructors.copy()
+
+    def _restore_modified_yaml_parameters(self) -> None:
+        yaml.Resolver.yaml_implicit_resolvers = self._modified_yaml_implicit_resolvers
+        yaml.SafeConstructor.yaml_constructors = self._modified_yaml_constructors
+
+    def load(self, stream: Union[Path, Any]) -> Any:
+        """Loads the YAML content.
+
+        Args:
+            stream: Any content or stream to parse.
+
+        Returns:
+            Parsed YAML.
+        """
+        self._restore_modified_yaml_parameters()
+        return super().load(stream)
+
+
+fix_yaml_loader()
+_parsers: Dict[Union[Text, List[Text]], YAMLParser] = {}
+ENV_VAR_REGEX = re.compile(r"\$\{[\S]+\}")
+
+
+def _get_yaml_parser(
+    reader_type: Union[Text, List[Text]] = "safe", replace_env_vars: bool = False
+) -> YAMLParser:
+    # we cannot reuse the existing parser in `_parsers` when `replace_env_vars` is `True`
+    # because the set of env variables might change
+    if replace_env_vars:
+        return YAMLParser(reader_type=reader_type, replace_env_vars=True)
+
+    if reader_type in _parsers:
+        return _parsers[reader_type]
+
+    new_parser = YAMLParser(reader_type=reader_type)
+    _parsers[reader_type] = new_parser
+    return new_parser
+
+
 def read_yaml(content: Text, reader_type: Union[Text, List[Text]] = "safe") -> Any:
     """Parses yaml from a text.
 
     Args:
         content: A text containing yaml content.
         reader_type: Reader type to use. By default "safe" will be used
+        replace_env_vars: Specifies if environment variables need to be replaced
 
     Raises:
         ruamel.yaml.parser.ParserError: If there was an error when parsing the YAML.
     """
-    fix_yaml_loader()
-
-    replace_environment_variables()
```

Comment:
In fact, the performance drop was also caused by this call being made every time we tried to read YAML. Multiple `SafeConstructor`s and implicit resolvers were being added, and every attempt to read YAML was making a situation even worse and making the whole process of env interpolation even slower.